### ISSUE TITLE
Option to grab season pack even if some episodes already meet cutoff

### ISSUE THREAD

**Issue by bbtufty** [Sonarr/Sonarr#6378]
### Is there an existing issue for this?

- [X] I have searched the existing open and closed issues

### Is your feature request related to a problem? Please describe

Currently, if you have a season with a mismatch of episodes (e.g. resolution or custom score), a season pack will not be grabbed unless it's an upgrade for all episodes. In these cases, you'd have to grab a season pack manually

### Describe the solution you'd like

It would be nice to have an option to grab if it's at least an upgrade for some (although likely with some checks to make sure this isn't a downgrade for any of the episodes). A checkbox in the profiles to allow this per-profile might be the neatest way to do this

### Describe alternatives you've considered

N/A

**Comment by thezoggy** [comment]
you can just manually grab it, which overrides any rejection. you then would also just need to manually import due to rejection.  or you could of course just delete the existing season (manage episodes/delete) before grabbing.

**Comment by bbtufty** [comment]
yep, this is what I currently do, I've got a note of seasons that are mixed episodes and then periodically check. But it seems like this could be implemented within sonarr itself rather than having to do it manually

**Comment by fryfrog** [comment]
Where would you draw the line? Would it be configurable or fixed?

The idea behind this is not to waste a bunch of download for small benefit.

As a work-a-round, you could use TRaSH's season pack regex to give a little bonus to them.

**Comment by bakerboy448** [comment]
Duplicate of https://github.com/Sonarr/Sonarr/issues/6242

**Comment by markus101** [comment]
From a previous discussion on this on Discord:

From Fryfrog:
> I would say if you're going to make this feature and give it an option, it should just be a 3 way toggle.
Off, Last 1-2 episodes, any episode.
Most people are asking it because of what you say, but I do see requests where people wish it'd do it for all of them.
And it should default to off

More recently he suggested making it percentage based, where it could either be:

Any - any episode is upgraded
X percentage (either 25/50/75 or 10/20/30/40/50/60/70/80/90)
All - same as we have now

Another option could be a slider, but I'm not sure that level of granularity is necessary.

On the import end this could be a little annoying because Sonarr won't import things which will leave the queue cluttered.

**Comment by bakerboy448** [comment]
Three options and a percentage base seems like a solid solution.  Seasonpackarr is doing the percentage thing as well

**Comment by markus101** [comment]
Something like:

- All (Default)
- Threshold (Shows separate number entry)
- Any (at least 1 episode)

Should there be a `Never` option to never grab them?

Where should this live, perhaps Media Management alongside the propers/repacks option?

**Comment by bakerboy448** [comment]
- All (Default) - Require a season pack to be a quality or custom format upgrade for all episodes
- Threshold - Require a season pack to be an upgrade for at least X percent of episodes. (additional help text showing calculated examples based on "common" season counts: 8, 12, 24)
- Any - Allow a season pack if it upgrades any episode. Note that this may result in stuck queue items.

Next to repack/propers would seem sane

**Comment by bbtufty** [comment]
This sounds like it would cover general use cases well, and mine in particular nicely

**Comment by akrigator** [comment]
Side effect of current implementation is that it also blocks to redownload updated torrents. For example  https://rutracker.org/  has single torrent for shows on air, and update it when new episode is available. So 
1. When you add show to sonarr,  it catch from rutracker some torrent S1E1-5
2. When episode 6 is released, the torrent is updated to S1E1-6, but sonarr filters out this torrent since it already has 5 of 6 episodes  with same quality.

**Comment by pariv** [comment]
The proposed solution for downloading season packs applies only to completed seasons, not to ongoing ones. Is that correct?

**Comment by sparky3387** [comment]
Not sure ive seen a partial season pack, but if set to any it would process it regardless as long as the conditions apply (this is an upgrade for any episode of the season)

**Comment by pariv** [comment]
The most popular trackers in my language name releases as "Show.S03E01-04.of.10.1080p" or similar and add episodes as they air. These are parsed as `FullSeason = false` (which is correct based on the title) and won't be processed by this season pack logic at all.

Note: #7877 describes similar tracker behavior and was closed as duplicate, but won't be solved by this either due to episode numbers in titles.

**Comment by sparky3387** [comment]
I suppose its true your example isn't a full season pack, if it was enabled would that not create a tonne of duplicate downloads? It appears there is no current check for if a season is full in existing code, does it currently process these at the moment, rejecting the release if you have any downloaded episodes? It would not take much to make this configurable, would be a fairly easy PR if you want me to guide you through this?

**Comment by pariv** [comment]
Yes, this could create multiple downloads as the pack gets updated with new episodes. However, if the user has configured a narrow quality upgrade range, the number of downloads wouldn't be significantly higher than downloading individual episodes.

This could be implemented as a user-configurable option with a clear warning about potential excessive storage usage.

Currently, when Sonarr sees a multi-episode release, it rejects it if ANY episode doesn't need an upgrade. The current patch changes this logic to be more flexible (Any/All/threshold), but only for releases where `FullSeason = true`. Partial season packs for ongoing/incomplete seasons are still processed with the old "all-or-nothing" logic.

**Comment by sparky3387** [comment]
I was going to start telling you what to put together in a PR, but looking on discord, I see

@stevietv 
sonarr will not automatically grab partial season packs, that is not changing

You might need to try new indexers, or alternatively you could write a helper script that pushes a release to see if its a show you have added in Sonarr, and then download the rejected release file directly, using node-red or something

**Comment by pariv** [comment]
Thanks for the clarification - I understand now.

It's disappointing since my main goal with Sonarr was automatic tracking of ongoing releases as new episodes air (with specifics of my regional trackers). But I appreciate the honest answer. I've already forked the project and made some changes for my needs, though they're not production-ready and will likely conflict with future releases.

I'll look into alternative approaches. Thanks for your time!

**Comment by fenihx** [comment]
Hi @pariv,
I have exactly the same issue as described here and in similar discussions. I work with Russian torrent trackers (like RuTracker, Kinozal) where season packs are frequently updated as new episodes are released.

Is there a way to get in touch with you directly to exchange ideas?

If you have any recommended workflows or are working on something similar, I’d be happy to help with testing or discussing possible solutions.

**Comment by Anvil5465** [comment]
@fenihx my understanding is that a pull request adding this feature has already been merged and will be in the release of Sonarr version 5: https://github.com/Sonarr/Sonarr/pull/7975

**Comment by fenihx** [comment]
@Anvil5465 I tried configuring this in v5-develop for testing. Maybe I'm doing something wrong, but it seems like this doesn't work with partial season packs. There's no automatic grabbing of partial seasons, even when episodesare missing

**Comment by stevietv** [comment]
the point of PR #7975 isn't to grab partial season packs. Its to grab season packs when some of the episodes already are downloaded.

In sonarr, a release needs to contain all episodes of a season to be considered a season pack. The way some russian trackers do it with updating the same release every time with more episodes isn't considered a season pack.

**Comment by pariv** [comment]
@fenihx, as  @Anvil5465 mentioned, Sonarr only considers releases with all season episodes as valid season packs. The reasoning seems to be outlined here: https://github.com/Sonarr/Sonarr/issues/5032#issuecomment-1135131376. The main concern is that, at the stage of parsing release titles, it's not always possible to know which episodes are actually present in the torrent. For Russian trackers, however, the episode info can often be extracted directly from the title format.

That said, the maintainers have decided not to deal with partial season packs at all. For now, I've made some code changes on my end so Sonarr accepts these types of releases, but robust support would require more parsing improvements to reliably handle our tracker formats.
